crypto/x509/x509_cmp.c

```diff
@@ -251,18 +251,21 @@ int X509_NAME_cmp(const X509_NAME *a, const X509_NAME *b)
         return -1;
 
     /* Ensure canonical encoding is present and up to date */
-    if (!a->canon_enc || a->modified) {
+    if (a->canon_enc == NULL || a->modified) {
         ret = i2d_X509_NAME((X509_NAME *)a, NULL);
         if (ret < 0)
             return -2;
     }
 
-    if (!b->canon_enc || b->modified) {
+    if (b->canon_enc == NULL || b->modified) {
         ret = i2d_X509_NAME((X509_NAME *)b, NULL);
         if (ret < 0)
             return -2;
     }
 
+    if (a->canon_enc == NULL || b->canon_enc == NULL)
```

Comment:
Looks like it could be possible inside x509_name_ex_i2d() for it to return 0.

Reply:
Again I refer you to a previous comment: https://github.com/openssl/openssl/pull/14589#discussion_r595859258

This is a comparison function, returning 0 means equality.  That's not a good return.